// common-confirmation-service/common-confirmation-service.js
// Общая библиотека Delivery
/* SQL Queries */
const SQL = {
  CONFIRMATION: {
    SQL_GET_CONFIRMATION_ENTRY: `SELECT * from pa_confirmation_service.verification_codes where request_id = $1`,     
    GET_ACTIVE_USER_REQUEST_ID : `
    SELECT request_id as "requestId" from pa_confirmation_service.verification_codes 
    where user_id = $1 and status = 'PENDING' and (NOW() - created) < INTERVAL '5 minutes'`, 
    SET_SUCCESS_REQUEST_ID : `
    UPDATE pa_confirmation_service.verification_codes
      set request_id = 'SUCCESS'
    where request_id = $1 and status = 'PENDING'`, 
    UPDATE_SEND_CONFIRM_CODE_RESULT : `
    UPDATE pa_confirmation_service.verification_codes
    set status = $2,  sended = now()
    where request_id = $1 
    RETURNING request_id`,
    SET_FAILED_REQUEST_ID : `
    UPDATE pa_confirmation_service.verification_codes
      set request_id = 'FAILED'
    where request_id = $1 and status = 'PENDING'`, 
    CREATE_CONFIRM_CODE : `INSERT INTO  pa_confirmation_service.verification_codes 
    ( user_id, confirmation_type, code) values 
    ( 
      $1, 
      $2, 
      FLOOR(RANDOM() * 90000 + 10000)) RETURNING request_id`   
  }
};

const MESSAGES = {
   RU : {    
      INPUT_VALIDATION_ERROR: 'Ошибка проверки входных данных',
   }
}

/* Экспорт */
module.exports = {
  SQL, MESSAGES
};
